cloud/services/availabilitysets/availabilitysets.go

```diff
@@ -0,0 +1,132 @@
+	azure "sigs.k8s.io/cluster-api-provider-azure/cloud"
+	"sigs.k8s.io/cluster-api-provider-azure/cloud/converters"
+	"sigs.k8s.io/cluster-api-provider-azure/cloud/services/resourceskus"
+	"sigs.k8s.io/cluster-api-provider-azure/util/tele"
+)
+
+// AvailabilitySetScope defines the scope interface for a availability sets service.
+type AvailabilitySetScope interface {
+	logr.Logger
+	azure.ClusterDescriber
+	AvailabilitySetSpecs() []azure.AvailabilitySetSpec
+}
+
+// Service provides operations on azure resources
+type Service struct {
+	Scope AvailabilitySetScope
+	Client
+	resourceSKUCache *resourceskus.Cache
+}
+
+// New creates a new availability sets service.
+func New(scope AvailabilitySetScope, skuCache *resourceskus.Cache) *Service {
+	return &Service{
+		Scope:            scope,
+		Client:           NewClient(scope),
+		resourceSKUCache: skuCache,
+	}
+}
+
+// Reconcile creates or updates availability sets
+func (s *Service) Reconcile(ctx context.Context) error {
+	ctx, span := tele.Tracer().Start(ctx, "availabilitysets.Service.Reconcile")
+	defer span.End()
+
+	asSku, err := s.resourceSKUCache.Get(ctx, string(compute.Aligned), resourceskus.AvailabilitySets)
+	if err != nil {
+		return errors.Wrapf(err, "failed to get availability sets sku")
+	}
+
+	maxFaultDomainCountStr, ok := asSku.GetCapability(resourceskus.MaximumPlatformFaultDomainCount)
+	if !ok {
+		return errors.Errorf("cannot find capability %s sku %s", resourceskus.MaximumPlatformFaultDomainCount, *asSku.Name)
+	}
+
+	maxFaultDomainCount, err := strconv.ParseUint(maxFaultDomainCountStr, 10, 32)
+	if err != nil {
+		return errors.Wrapf(err, "failed to determine max fault domain count")
+	}
+
+	for _, asSpec := range s.Scope.AvailabilitySetSpecs() {
+		s.Scope.V(2).Info("creating availability set", "availability set", asSpec.Name)
+
+		faultDomainCount := asSpec.DesiredFaultDomainCount
+		if asSpec.DesiredFaultDomainCount > int32(maxFaultDomainCount) {
+			s.Scope.V(2).Info("desired fault domain count is greater than max fault domain count, will use the max fault domain count",
+				"desired fault domain count", asSpec.DesiredFaultDomainCount, "max fault domain count", maxFaultDomainCount)
+			faultDomainCount = int32(maxFaultDomainCount)
+		}
+
+		asParams := compute.AvailabilitySet{
+			Sku: &compute.Sku{
+				Name: pointer.StringPtr(string(compute.Aligned)),
+			},
+			AvailabilitySetProperties: &compute.AvailabilitySetProperties{
+				PlatformFaultDomainCount: pointer.Int32Ptr(faultDomainCount),
+			},
+			Tags: converters.TagsToMap(infrav1.Build(infrav1.BuildParams{
+				ClusterName: s.Scope.ClusterName(),
+				Lifecycle:   infrav1.ResourceLifecycleOwned,
+				Name:        to.StringPtr(asSpec.Name),
+				Role:        to.StringPtr(infrav1.CommonRole),
+				Additional:  s.Scope.AdditionalTags(),
+			})),
+			Location: pointer.StringPtr(s.Scope.Location()),
+		}
+
+		_, err := s.Client.CreateOrUpdate(ctx, s.Scope.ResourceGroup(), asSpec.Name, asParams)
```

Comment:
Just a note: with this change, an availability set will be created for all existing clusters but the VMs won't be added to it since they already exist. This won't cause any functional issues but the AS will be unused, which might be a bit confusing to users. However, any new control plane VMs (eg. created as part of a k8s upgrade or a scale up) would added to the set. Might worth calling that out in the PR description.

Reply:
Updated the `Special notes for your reviewer` section to add this note.

Reply:
maybe add it to the release notes?

Reply:
Updated release notes as well, Thanks.